### PR TITLE
Issue #54 - Single Item expanded at a time

### PR DESCRIPTION
**Pull request by Kwood22**
## Purpose:
To re-introduce the functionality to be able to toggle between having multiple items open at once or to have a single item open at a time. I have also added the functionality to allow a parent cell to be initially expanded. I have also updated the readme to include mention of this functionality as well as screenshots of the examples.

*Note that the default behavior is to have multiple items open at once*.

## Solution Implemented:
1. Inspiration for solution: [Release 1.2.6](https://github.com/Vkt0r/AccordionSwift/releases/tag/v1.2.6)
2. Introduced an enum for the number of expanded parent cells ie. `Single` or `Multiple`
3. Introduced a private instance variable to keep track of the expanded cell to be able to close it when another cell is selected

#### Contract updated: DataSourceProvider
```swift
DataSourceProvider(
    dataSource: DataSource,
    parentCellConfig: ParentCellConfig,
    childCellConfig: ChildCellConfig,
    didSelectParentAtIndexPath: DidSelectParentAtIndexPathClosure?,
    didSelectChildAtIndexPath: DidSelectChildAtIndexPathClosure?,
    heightForParentCellAtIndexPath: HeightForParentAtIndexPathClosure?,
    heightForChildCellAtIndexPath: HeightForChildAtIndexPathClosure?,
    scrollViewDidScroll: ScrollViewDidScrollClosure?,
    numberOfExpandedParentCells: NumberOfExpandedParentCells = .multiple // or .single
)
```

## Issue:  
[Issue 54](https://github.com/Vkt0r/AccordionSwift/issues/54)

#### Default behaviour:
screenshot

#### Only single parent allowed open at a time:
screenshot

#### Only single parent allowed open at a time but with initial parent expanded
screenshot

### CODE REVIEW

File: Example/Views/AccordionTableViewController.swift [hunk 1]
```diff
@@ -32,106 +33,108 @@ class AccordionTableViewController: UITableViewController {
...
-            didSelectChildAtIndexPath: didSelectChildCell,
-            scrollViewDidScroll: scrollViewDidScroll
+                dataSource: dataSource,
+                parentCellConfig: parentCellConfig,
+                childCellConfig: childCellConfig,
+                didSelectParentAtIndexPath: didSelectParentCell,
+                didSelectChildAtIndexPath: didSelectChildCell,
+                scrollViewDidScroll: scrollViewDidScroll,
+                // Configure DataSourceProvider to have only one parent expanded at a time
+                numberExpandedParentCells: .single
```
Comment (DewaldDeJager) [inline Example/Views/AccordionTableViewController.swift]: Consider naming this `numberOfExpandedParentCells` for better readability.

File: README.md [hunk 2]
```diff
@@ -121,10 +121,20 @@ public protocol CellViewConfigType {
 }
 ```
 
-The another step is define the `DataSourceProvider` in charge of handle the data source and the `CellViewConfig` for each cell.
+Another step is to define the `DataSourceProvider` in charge of handling the data source and the `CellViewConfig` for each cell. 
+The `DataSourceProvider` exposes the `NumberOfExpandedParentCells` attribute which can be used to change the behavior of the accordion to only have a `single` item open at once or to have `multiple` items open at any given time. Take note that the default behavior is to have multiple items open.
```
Comment (DewaldDeJager) [inline README.md]: Lowercase `n` in `NumberOfExpandedParentCells` and remember to update this if you do decide to rename it.

**Comment by Kwood22** [comment]
@Vkt0r Please could you review the pull request. I would greatly appreciate any feedback on this.

**Comment by DewaldDeJager** [comment]
> @Kwood22 @DewaldDeJager Thanks for this PR and the review. I'll take a look today, sorry for the delay.

@Vkt0r Any update on this?

**Comment by Vkt0r** [comment]
@DewaldDeJager @Kwood22 Sorry for the delay again I've been very busy. I would check it today definitely, stay tuned!

File: Example/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    dataSource: dataSource,
+    parentCellConfig: parentCellConfig,
+    childCellConfig: childCellConfig,
+    didSelectParentAtIndexPath: didSelectParentCell,
+    didSelectChildAtIndexPath: didSelectChildCell,
+    scrollViewDidScroll: scrollViewDidScroll,
+    // Configure DataSourceProvider to have only one parent expanded at a time
+    numberOfExpandedParentCells: .single, // .multiple
+    // Expand the 4th element initially
+    expandParentAtIndex: 3
```
Comment (Vkt0r) [inline Example/README.md]: We expose each `Parent` its current `state` in which it can be set the `collapsed` and `expanded` state. I don't see why we need to add a new verbose variable `expandParentAtIndex` in the `DataSourceProvider`.

File: Source/DataSource.swift [hunk 4]
```diff
@@ -42,29 +42,35 @@ public struct DataSource<Item: ParentType> {
...
     /// Inserts the item at the specified row and section.
     ///
     /// - Parameters:
     ///   - item: The item to insert.
     ///   - row:  The row index of the item.
     ///   - section: The section index of the item.
     public mutating func insert(item: Item, atRow row: Int, inSection section: Int) {
-        guard section < numberOfSections() else { return }
-        guard row <= numberOfItems(inSection: section) else { return }
+        guard section < numberOfSections() else {
```
Comment (Vkt0r) [inline Source/DataSource.swift]: It would be very helpful to keep the current style guide of the repository, it makes the review more complicated as there are a lot of unnecessary changes. Can you please keep it in the original form?

File: Source/DataSourceProvider.swift [hunk 5]
```diff
@@ -3,218 +3,339 @@
...
         self.heightForParentCellAtIndexPath = heightForParentCellAtIndexPath
         self.heightForChildCellAtIndexPath = heightForChildCellAtIndexPath
         self.scrollViewDidScroll = scrollViewDidScroll
+        self.numberOfExpandedParentCells = numberOfExpandedParentCells
+
+        var mutableDataSource = dataSource
+        let numberOfParentCells = mutableDataSource.numberOfParents()
+
+        if numberOfParentCells == 0 {
+            os_log("The data source does not contain any parents", type: .error)
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: I think assertions would make more sense here instead of `os_log`.

File: Source/DataSourceProvider.swift [hunk 6]
```diff
@@ -3,218 +3,339 @@
...
+        tableView.insertRows(at: indexPaths, with: .fade)
+        dataSource.toggleParentCell(toState: .expanded, inSection: parent.indexPath.section, atIndex: parent.index)
+    }
+
+// Collapse the parent cell and it's children
+//
+// - Parameters:
+//   - parent: The actual parent cell to be expanded
+    private func collapse(parent: ParentCell, _ tableView: UITableView) {
+        let numberOfChildren = dataSource.item(atRow: parent.index, inSection: parent.indexPath.section)?.children.count ?? 0
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: Can we encapsulate `parent.indexPath.section` & `parent.index` to avoid repetition ?

File: Source/DataSourceProvider.swift [hunk 7]
```diff
@@ -3,218 +3,339 @@
...
+            expand(parent: selectedParentCell, tableView)
+        }
+    }
+
+// Expand the parent cell and it's children
+//
+// - Parameters:
+//   - parent: The actual parent cell to be expanded
+    private func expand(parent: ParentCell, _ tableView: UITableView) {
+        let numberOfChildren = dataSource.item(atRow: parent.index, inSection: parent.indexPath.section)?.children.count ?? 0
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: Can we encapsulate parent.indexPath.section & parent.index to avoid repetition ?

File: Source/DataSourceProvider.swift [hunk 8]
```diff
@@ -3,218 +3,339 @@
...
+                return parent.index
+            case .multiple:
+                // Make use of indexPath if multiple parents can be expanded as indexPath.row will be up to date
+                return parent.indexPath.row
+            }
+        }()
+        return (1...numberOfChildren).map {
+            offset -> IndexPath in
+            IndexPath(row: startPosition + offset, section: parent.indexPath.section)
+        }
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: There is some misalignment here I think.

**Comment by Vkt0r** [comment]
@Kwood22 Thanks again for taking the time to create this. Looks good in general 👍. I've got a few comments for you.

**Comment by Kwood22** [comment]
> hanks again for taking the time to create this. Looks good in general 👍. I've got a few comments for you.

Thank you for the feedback. I will get to answering the comments soon.

File: Source/DataSource.swift (hunk 4 above)
Comment (Kwood22) [reply Source/DataSource.swift]: Could you please advise what the style guide for the repository is. I will then go reformat the files. Sorry about this.

**Comment by Kwood22** [comment]
@Vkt0r Please see the updated changes.

File: Source/DataSource.swift (hunk 4 above)
Comment (Vkt0r) [reply Source/DataSource.swift]: Oh no problem at all, actually is not a final style guide is defined. The issue is with changes in style the PR grows in size and makes it more difficult to review the important parts 😅. If we want to update the style guide we should create a separate PR only for that in the future.

**Comment by Vkt0r** [comment]
@Kwood22 I'll do. Thanks 👍.

**Comment by Kwood22** [comment]
@Vkt0r Have you had a chance to review the changes?

**Comment by Vkt0r** [comment]
@Kwood22 I'll today don't worry.

File: Source/DataSourceProvider.swift [hunk 9]
```diff
@@ -3,218 +3,342 @@
 //  AccordionSwift
 //
 //  Created by Victor Sigler Lopez on 7/3/18.
+//  Updated by Kyle Wood on 15/10/19.
 //  Copyright © 2018 Victor Sigler. All rights reserved.
 //
 
 import Foundation
+import os.log
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: I don't think we need this anymore isn't ?

File: Source/DataSourceProvider.swift [hunk 10]
```diff
@@ -3,218 +3,342 @@
...
+// MARK: - Private Methods
+
+// Update the cells of the table based on the selected parent cell
+//
+// - Parameters:
+//   - tableView: The UITableView to update
+//   - item: The DataSource item that was selected
+//   - currentPosition: The current position in the data source
+//   - indexPaths: The last IndexPath of the new cells expanded
+//   - parentIndex: The index of the parent item selected
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: There is some formatting issue here.

File: Source/DataSourceProvider.swift [hunk 11]
```diff
@@ -3,218 +3,342 @@
...
+            // Scroll the new cells expanded in case of be outside the UITableView CGRect
+            scrollCellIfNeeded(atIndexPath: lastCellIndexPath, tableView)
+        }
+    }
+
+// Toggle the state of the selected parent cell between expanded and collapsed
+//
+// - Parameters:
+//   - currentState: The current state of the selected parent
+//   - selectedParentCell: The actual cell selected
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: Same formatting issue.

File: Source/DataSourceProvider.swift [hunk 12]
```diff
@@ -3,218 +3,342 @@
...
+        toggle(selectedParentCell, withState: item.state, dataSourceIndex: parentIndex, tableView)
+        tableView.endUpdates()
+
+        // If the cells were expanded then we verify if they are inside the CGRect
+        if item.state == .expanded {
+            let lastCellIndexPath = IndexPath(item: indexPath.item + numberOfChildren, section: indexPath.section)
+            // Scroll the new cells expanded in case of be outside the UITableView CGRect
+            scrollCellIfNeeded(atIndexPath: lastCellIndexPath, tableView)
+        }
+    }
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: This looks very good 👍.

File: Source/DataSourceProvider.swift [hunk 13]
```diff
@@ -3,218 +3,342 @@
...
+            expand(parent: selectedParentCell, dataSourceIndex: dataSourceIndex, tableView)
+        }
+    }
+
+// Expand the parent cell and it's children
+//
+// - Parameters:
+//   - parent: The actual parent cell to be expanded
+//   - dataSourceIndex: The index of the parent in the dataSource
+//   - tableView: The tableView to insert the children cells into
```
Comment (Vkt0r) [inline Source/DataSourceProvider.swift]: Formatting too!

File: Source/DataSource.swift [hunk 14]
```diff
@@ -85,77 +93,138 @@ public struct DataSource<Item: ParentType> {
...
+        parents[index].state = state
+
+        sections[section].total += {
+            var childrenCount = parents[index].children.count
+            if state == .collapsed {
+                childrenCount *= -1
+            }
+            return childrenCount
+        }()
+    }
```
Comment (Vkt0r) [inline Source/DataSource.swift]: There is any specific advantage of using a closure here?

File: Source/DataSource.swift [hunk 15]
```diff
@@ -85,77 +93,138 @@ public struct DataSource<Item: ParentType> {
...
+
+    public func indexOfFirstExpandedParent() -> IndexPath? {
+        for section in (0..<self.numberOfSections()) {
+            if let parents = items(inSection: section) {
+                for parentIndex in parents.indices {
+                    if parents[parentIndex].state == .expanded {
+                        return IndexPath(item: parentIndex, section: section)
```
Comment (Vkt0r) [inline Source/DataSource.swift]: What do you think if we remove a little of imperative programming using something like this here IMO:
```swift
 return parents
      .firstIndex(where: { $0.state == .expanded })
      .map { IndexPath(item: $0, section: section) }
```

**Comment by Vkt0r** [comment]
@Kwood22 A posted a few comments about it. With these changes, we're very close to merging this. Great work doing this !

File: Source/DataSource.swift (hunk 15 above)
Comment (Kwood22) [reply Source/DataSource.swift]: I have tried to implement the changes you suggested, I encountered a couple of issues so I will walk you through my thought process.

#### Attempt 1: 
I replaced the inner for loop with the firstIndex. I found that I needed to introduce a variable `indexPath` to allow the first for loop to short circuit if a parent that is expanded is found. I also didn't like the mix of imperial and functional programming. That lead me to attempt2.
```swift
public func indexOfFirstExpandedParent() -> IndexPath? {
   var indexPath: IndexPath? = nil
   for section in (0..<self.numberOfSections()) {
      if let parents = items(inSection: section) {
          return parents.firstIndex(where: { $0.state == .expanded })
                        .map { indexPath = IndexPath(item: $0, section: section) }
      }
      
      if indexPath == nil {
        return indexPath
      }
  }
  return nil
}
```

#### Attempt 2:
I decided to go with an all functional approach. I encountered the following issues:
- In the below code the reference to the section in which the parent belongs gets lost in the flat map and thus the indexpath produced is incorrect.


```swift
public func indexOfFirstExpandedParent() -> IndexPath? {
    (0..<self.numberOfSections())
            .map { section in items(inSection: section) }
            .flatMap{ $0 ?? []}
            .firstIndex(where: { $0.state == .expanded })
            .map { IndexPath(item: $0, section: // Lost in flatmap) }
}
```

I then decided to try wrap the parents and section in a tuple. Please see the below signatures:
```
map:     [Int] -> ([Item]?, Int)
flatMap: [([Item]?, Int)] -> [(Item, Int)] // I had trouble getting this correct conversion working correctly in a clean way
first:   [(Item, Int)] -> (Item, Int)      // (where state == expanded )
map:     (Item, Int) -> IndexPath?
```
I was not able to get the flatmap to return what I wanted, I thus have left the implementation of Attempt 1. 
Please feel free to point out if I missed something obvious, or if you have a better solution.

Comment (Kwood22) [reply Source/DataSource.swift]: Thank you for the response. I agree that we don't need the index path variable, I have removed it from the below code.  Does this look more correct?
```swift
public func indexOfFirstExpandedParent() -> IndexPath? {
    // Loop through each section if there are multiple
    for section in (0..<self.numberOfSections()) {
        if let parents = items(inSection: section) {
            // Find parent that is expanded
            let indexPath = parents.firstIndex(where: { $0.state == .expanded })
                                   .map { IndexPath(item: $0, section: section) }

            // Return the index path if a parent is found to be expanded
            if indexPath != nil {
                return indexPath
            }
        }
    }
    return nil
}
```

Please see the changes to the PR, I have updated the code to the above.

Comment (Vkt0r) [reply Source/DataSource.swift]: Yep, that looks very good!

File: Source/DataSource.swift [hunk 16]
```diff
@@ -93,129 +93,101 @@ public struct DataSource<Item: ParentType> {
...
     public func numberOfExpandedParents() -> Int {
         var count = 0
         for section in (0..<self.numberOfSections()) {
             if let parents = items(inSection: section) {
-                for parentIndex in parents.indices {
-                    if parents[parentIndex].state == .expanded {
-                        count += 1
-                    }
-                }
+                count += parents.filter { $0.state == .expanded }.count
```
Comment (Vkt0r) [inline Source/DataSource.swift]: 👍.

**Comment by Kwood22** [comment]
@Vkt0r  Thank you for all the comments and commitment! 
Will you be looking to release a new version soon?

**Comment by Vkt0r** [comment]
Yes, I'll during the weekend.

**Comment by Kwood22** [comment]
> Yes, I'll during the weekend.

@Vkt0r  That will be great thank you!!
